kotlinx-coroutines-core/common/src/internal/SegmentQueue.kt

```diff
@@ -0,0 +1,186 @@
+                    newTail
+                } else {
+                    cur.next.value!!
+                }
+            }
+            cur = curNext
+        }
+        if (cur.id != id) return null
+        return cur
+    }
+
+    /**
+     * Invokes [getSegment] and replaces [first] with the result if its [id] is greater.
+     */
+    protected fun getSegmentAndMoveFirst(startFrom: S?, id: Long): S? {
+        if (startFrom !== null && startFrom.id == id) return startFrom
+        val s = getSegment(startFrom, id) ?: return null
+        moveHeadForward(s)
+        return s
+    }
+
+    /**
+     * Updates [_head] to the specified segment
+     * if its `id` is greater.
+     */
+    private fun moveHeadForward(new: S) {
+        while (true) {
+            val cur = first!!
+            if (cur.id > new.id) return
+            if (_head.compareAndSet(cur, new)) {
+                new.prev.value = null
+                return
+            }
+        }
+    }
+
+    /**
+     * Updates [_tail] to the specified segment
+     * if its `id` is greater.
+     */
+    private fun moveTailForward(new: S) {
+        while (true) {
+            val cur = last
+            if (cur !== null && cur.id > new.id) return
+            if (_tail.compareAndSet(cur, new)) return
+        }
+    }
+}
+
+/**
+ * Each segment in [SegmentQueue] has a unique id and is created by [SegmentQueue.newSegment].
+ * Essentially, this is a node in the Michael-Scott queue algorithm, but with
+ * maintaining [prev] pointer for efficient [remove] implementation.
+ */
+internal abstract class Segment<S: Segment<S>>(val id: Long, prev: S?) {
+    // Pointer to the next segment, updates similarly to the Michael-Scott queue algorithm.
+    val next = atomic<S?>(null)
+    // Pointer to the previous segment, updates in [remove] function.
+    val prev = atomic<S?>(null)
+
+    /**
+     * Returns `true` if this segment is logically removed from the queue.
+     * The [remove] function should be called right after it becomes logically removed.
+     */
+    abstract val removed: Boolean
+
+    init {
+        this.prev.value = prev
+    }
+
+    /**
+     * Removes this segment physically from the segment queue. The segment should be
+     * logically removed (so [removed] returns `true`) at the point of invocation.
+     */
+    fun remove() {
+        check(removed) { " The segment should be logically removed at first "}
+        // Read `next` and `prev` pointers.
+        val next = this.next.value ?: return // tail cannot be removed
+        val prev = prev.value ?: return // head cannot be removed
+        // Link `next` and `prev`.
+        next.movePrevToLeft(prev)
+        prev.movePrevNextToRight(next)
+        // Check whether `prev` and `next` are still in the queue
+        // and help with removing them if needed.
+        if (prev.removed)
+            prev.remove()
+        if (next.removed)
+            next.remove()
+    }
```

Comment:
Commenting out these four last lines breaks no test. There are not enough concurrent stress-test to detect any potential bugs in this code.